docs/better-burgers/index.md

```diff
@@ -0,0 +1,345 @@
+
+```reason
+let toPrice = ({onions, cheese, tomato, bacon, lettuce: _}) => {
+  let toppingCost = /* write me */;
+
+  15.  // base cost
+  +. toppingCost(onions, 0.2)
+  +. toppingCost(cheese, 0.1)
+  /* some stuff involving tomato and bacon */;
+};
+```
+
+<b>4.</b> Right now, the `Burger.toEmoji` function shows more emojis than
+absolutely necessary. Refactor the `multiple` inner function in `Burger.toEmoji`
+so that it exhibits the following behavior:
+
+| Burger | Expected output |
+| ------ | --------------- |
+| `Burger({lettuce: true, onions: 1, cheese: 1})` | `🍔{🥬,🧅,🧀}` |
+| `Burger({lettuce: true, onions: 0, cheese: 0})` | `🍔{🥬}` |
+
+::: details Hint
+
+Use a switch expression.
+
+:::
+
+## Overview
+
+- Record types are like `Js.t` objects but their fields must be explicitly
+  defined
+  - Records and `Js.t` objects are both JavaScript objects during runtime
+- You can get the fields out of a record using destructuring and pattern
+  matching:
+  - `let` destructuring:
+    ```reason
+    let {a, b, c} = record;
+    ```
+  - Switch expression branch pattern matching:
+    ```reason
+    switch (record) {
+    | {a: "foo", b: "bar", c: 42} => "Magnifique!"
+    | {a, b, c} => a ++ b ++ string_of_int(c)
+    };
+    ```
+- It's common practice to group related types and functions into a submodule
+- Try not to ignore record fields when pattern matching on records. Instead of
+  ```reason
+  | {a, b, _} =>
+  ```
+
+  Prefer
+
+  ```reason
+  | {a, b, c: _, d: _} =>
+  ```
+  Assuming `c` and `d` aren't used inside the branch.
+- Try not to pattern match on tuples of more than 2 elements because it tends to
+  be hard to read
+
+## Solutions
+
+<b>1.</b> The `Item.Sandwich` submodule should look something like this:
+
+<<< Item.re#sandwich-module
+
+You also need to refactor `Item.toPrice` and `Item.toEmoji` functions to use the
+new functions in `Item.Sandwich`.
+
+<b>2.</b> After adding `tomatoes` and `bacon` fields to `Item.Burger.t`, the
+changes to `Item.Burger.toEmoji` are fairly mechanical, so let's focus on
+`Item.Burger.toPrice`:
+
+<<< Item.re#to-price-tomato-bacon
+
+Note that we keep `lettuce: _` at the end of the pattern match since the value
+of `lettuce` isn't ever used.
+
+<b>3.</b> The `toppingCost` helper function inside `Burger.toPrice` should look
+something like this:
+
+<<< Item.re#to-price-topping-cost
+
+<b>4.</b> After refactoring the `multiple` helper function inside
+`Burger.toEmoji` to avoid showing unnecessary emojis, it should look something
+like this:
+
+<<< Item.re#to-emoji-multiple
+
+Since the body of this function only consists of a switch expression, you can
+actually refacor the `multiple` function to use the `fun` syntax:
+
+<<< Item.re#to-emoji-multiple-fun
+
+This might look a little strange, but it works because a two-argument function
+in OCaml is, in reality, a one-argument function that returns another function
+that accepts a single argument[^2]. Manually adding a type annotation to the
```

Comment:
Maybe we should say something like "in OCaml, all functions take exactly one argument".

Reply:
Fixed in https://github.com/melange-re/melange-for-react-devs/pull/21/commits/4f44abc0eb2808354fca6e0c3718054cded994e4